nsfc-plugin/src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java

```diff
@@ -87,25 +167,139 @@ public void removeInspectionPort(InspectionPortElement inspectionPort)
+
+                if (portPairGroup.getPortPairs().size() > 0) {
+                    PortPairGroup ppgUpdate = Builders.portPairGroup().portPairs(portPairGroup.getPortPairs()).build();
+                    this.osCalls.updatePortPairGroup(portPairGroup.getId(), ppgUpdate);
+                } else {
+                    PortChain portChain = this.utils.fetchContainingPortChain(portPairGroup.getId());
+
+                    if (portChain != null) {
+                        List<String> ppgIds = portChain.getPortPairGroups();
+                        ppgIds.remove(portPairGroup.getId());
+
+                        // service function chain with with no port pair should be allowed to exist?
+                        PortChain portChainUpdate = Builders.portChain().portPairGroups(ppgIds).build();
+                        this.osCalls.updatePortChain(portChain.getId(), portChainUpdate);
+                    }
+                    this.osCalls.deletePortPairGroup(portPairGroup.getId());
+                }
+            }
+
+            this.osCalls.deletePortPair(portPair.getId());
+        } else {
+            LOG.warn("Attempt to remove nonexistent Inspection Port for ingress {} and egress {}",
+                    inspectionPort.getIngressPort(), inspectionPort.getEgressPort());
+        }
     }
 
     // Inspection Hooks methods
     @Override
-    public String installInspectionHook(NetworkElement inspectedPort, InspectionPortElement inspectionPort, Long tag,
-            TagEncapsulationType encType, Long order, FailurePolicyType failurePolicyType)
+    public String installInspectionHook(NetworkElement inspectedPortElement,
+                                        InspectionPortElement inspectionPortElement, Long tag,
+                                        TagEncapsulationType encType, Long order,
+                                        FailurePolicyType failurePolicyType)
             throws NetworkPortNotFoundException, Exception {
 
-        return null; // TODO (Dmitry) Implement
+        checkArgument(inspectedPortElement != null && inspectedPortElement.getElementId() != null,
+                      "null passed for %s !", "Service Function Chain");
+        checkArgument(inspectionPortElement != null && inspectionPortElement.getElementId() != null,
+                      "null passed for %s !", "Service Function Chain");
+
+        if (inspectedPortElement.getPortIPs() == null || inspectedPortElement.getPortIPs().size() == 0) {
+            String msg = String.format("Inspected port %s has no address to protect!", inspectedPortElement.getElementId());
+            LOG.error(msg);
+            throw new IllegalStateException(msg);
+        }
+
+        LOG.info("Installing Inspection Hook for (Inspected Port {} ; Inspection Port {}):",
+                inspectedPortElement, inspectionPortElement);
+
+        PortChain portChain = this.osCalls.getPortChain(inspectionPortElement.getElementId());
```

Comment:
we dont need to check `portChain.getId()!=null` since its coming from openstack and is a required field there.